File: src/main/java/org/panteleyev/money/app/TransactionTableView.java
/*
 Copyright (c) Petr Panteleyev. All rights reserved.
 Licensed under the BSD license. See LICENSE file in the project root for full license information.
 */
package org.panteleyev.money.app;

import javafx.application.Platform;
import javafx.beans.property.ReadOnlyIntegerProperty;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;
import javafx.collections.WeakListChangeListener;
import javafx.collections.transformation.FilteredList;
import javafx.scene.control.Alert;
import javafx.scene.control.ButtonType;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.SelectionMode;
import javafx.scene.control.SeparatorMenuItem;
import javafx.scene.control.TableCell;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.stage.FileChooser;
import javafx.util.Callback;
import org.panteleyev.fx.PredicateProperty;
import org.panteleyev.money.app.cells.TransactionAccountRequestSumCell;
import org.panteleyev.money.app.cells.TransactionCheckCell;
import org.panteleyev.money.app.cells.TransactionCommentCell;
import org.panteleyev.money.app.cells.TransactionContactCell;
import org.panteleyev.money.app.cells.TransactionCreditedAccountCell;
import org.panteleyev.money.app.cells.TransactionDayCell;
import org.panteleyev.money.app.cells.TransactionDebitedAccountCell;
import org.panteleyev.money.app.cells.TransactionRow;
import org.panteleyev.money.app.cells.TransactionSumCell;
import org.panteleyev.money.app.cells.TransactionTypeCell;
import org.panteleyev.money.app.details.TransactionDetailsDialog;
import org.panteleyev.money.app.options.Options;
import org.panteleyev.money.model.Account;
import org.panteleyev.money.model.Category;
import org.panteleyev.money.model.Contact;
import org.panteleyev.money.model.Transaction;
import org.panteleyev.money.model.TransactionDetail;
import org.panteleyev.money.persistence.MoneyDAO;
import org.panteleyev.money.xml.Export;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.math.BigDecimal;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import static org.panteleyev.fx.FxUtils.fxString;
import static org.panteleyev.fx.MenuFactory.menuItem;
import static org.panteleyev.fx.TableColumnBuilder.tableObjectColumn;
import static org.panteleyev.money.app.Constants.ELLIPSIS;
import static org.panteleyev.money.app.Constants.FILTER_ALL_FILES;
import static org.panteleyev.money.app.Constants.FILTER_XML_FILES;
import static org.panteleyev.money.app.MainWindowController.RB;
import static org.panteleyev.money.app.options.Options.options;
import static org.panteleyev.money.persistence.DataCache.cache;
import static org.panteleyev.money.persistence.MoneyDAO.getDao;

public class TransactionTableView extends TableView<Transaction> {
    public interface TransactionDetailsCallback {
        void handleTransactionDetails(Transaction transaction, List<TransactionDetail> details);
    }

    public enum Mode {
        SUMMARY(false),
        STATEMENT(true),
        QUERY(true),
        ACCOUNT(true);

        private final boolean fullDate;

        Mode(boolean fullDate) {
            this.fullDate = fullDate;
        }

        public boolean isFullDate() {
            return fullDate;
        }
    }

    private final Mode mode;

    private BiConsumer<List<Transaction>, Boolean> checkTransactionConsumer = (x, y) -> { };
    private final TransactionDetailsCallback transactionDetailsCallback;

    private final Consumer<Transaction> transactionAddedCallback;
    private final Consumer<Transaction> transactionUpdatedCallback;

    @SuppressWarnings("FieldCanBeLocal")
    private final ListChangeListener<Contact> contactChangeLister = change -> Platform.runLater(this::redraw);

    @SuppressWarnings("FieldCanBeLocal")
    private final ListChangeListener<Category> categoryChangeLister = change -> Platform.runLater(this::redraw);

    // Transaction filter
    private final PredicateProperty<Transaction> transactionPredicateProperty = new PredicateProperty<>(x -> false);

    private final FilteredList<Transaction> filteredList = new FilteredList<>(cache().getTransactions());

    // List size property
    private final SimpleIntegerProperty listSizeProperty = new SimpleIntegerProperty(0);

    TransactionTableView(Mode mode) {
        this(mode, null, null, x -> {}, x -> {});
    }

    TransactionTableView(Account account) {
        this(Mode.ACCOUNT, account, null, x -> {}, x -> {});
    }

    TransactionTableView(Mode mode,
                         TransactionDetailsCallback transactionDetailsCallback,
                         Consumer<Transaction> transactionAddedCallback,
                         Consumer<Transaction> transactionUpdatedCallback)
    {
        this(mode, null, transactionDetailsCallback, transactionAddedCallback, transactionUpdatedCallback);
    }

    TransactionTableView(Mode mode, Account account,
                         TransactionDetailsCallback transactionDetailsCallback,
                         Consumer<Transaction> transactionAddedCallback,
                         Consumer<Transaction> transactionUpdatedCallback)
    {
        this.mode = mode;
        if (mode == Mode.ACCOUNT && account == null) {
            throw new IllegalArgumentException("Account cannot be null when mode = ACCOUNT");
        }

        this.transactionDetailsCallback = transactionDetailsCallback;
        this.transactionAddedCallback = transactionAddedCallback;
        this.transactionUpdatedCallback = transactionUpdatedCallback;

        setRowFactory(x -> new TransactionRow());

        getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);

        var w = widthProperty().subtract(20);
        var dayComparator = mode.isFullDate() ? MoneyDAO.COMPARE_TRANSACTION_BY_DATE : MoneyDAO.COMPARE_TRANSACTION_BY_DAY;

        Callback<TableColumn<Transaction, Transaction>, TableCell<Transaction, Transaction>> sumCellFactory =
            mode == Mode.ACCOUNT ?
                x -> new TransactionAccountRequestSumCell(account) :
                x -> new TransactionSumCell();

        getColumns().setAll(List.of(
            tableObjectColumn(fxString(RB, "Day"), b ->
                b.withCellFactory(x -> new TransactionDayCell(mode.isFullDate()))
                    .withComparator(dayComparator)
                    .withWidthBinding(w.multiply(0.05))),
            tableObjectColumn(fxString(RB, "Type"), b ->
                b.withCellFactory(x -> new TransactionTypeCell())
                    .withComparator(Comparator.comparingInt((Transaction t) -> t.type().ordinal())
                        .thenComparing(dayComparator))
                    .withWidthBinding(w.multiply(0.1))),
            tableObjectColumn(fxString(RB, "column.Account.Debited"), b ->
                b.withCellFactory(x -> new TransactionDebitedAccountCell())
                    .withComparator(Comparator.comparing(Transaction::accountDebitedUuid)
                        .thenComparing(dayComparator))
                    .withWidthBinding(w.multiply(0.1))),
            tableObjectColumn(fxString(RB, "column.Account.Credited"), b ->
                b.withCellFactory(x -> new TransactionCreditedAccountCell()).withWidthBinding(w.multiply(0.1))),
            tableObjectColumn(fxString(RB, "Counterparty"), b ->
                b.withCellFactory(x -> new TransactionContactCell()).withWidthBinding(w.multiply(0.2))),
            tableObjectColumn(fxString(RB, "Comment"), b ->
                b.withCellFactory(x -> new TransactionCommentCell()).withWidthBinding(w.multiply(0.35))),
            tableObjectColumn(fxString(RB, "Sum"), b ->
                b.withCellFactory(sumCellFactory)
                    .withComparator(Comparator.comparing(Transaction::getSignedAmount))
                    .withWidthBinding(w.multiply(0.05))),
            tableObjectColumn("", b ->
                b.withCellFactory(x -> new TransactionCheckCell()).withWidthBinding(w.multiply(0.05)))
        ));

        getSortOrder().add(getColumns().get(0));

        createContextMenu();

        cache().getContacts().addListener(new WeakListChangeListener<>(contactChangeLister));
        cache().getCategories().addListener(new WeakListChangeListener<>(categoryChangeLister));

        filteredList.predicateProperty().bind(transactionPredicateProperty);

        var sortedList = filteredList.sorted();
        sortedList.comparatorProperty().bind(comparatorProperty());
        setItems(sortedList);
    }

    ReadOnlyIntegerProperty listSizeProperty() {
        return listSizeProperty;
    }

    private void createContextMenu() {
        var newMenuItem = menuItem(fxString(RB, "Add", ELLIPSIS), event -> onNewTransaction());
        var editMenuItem = menuItem(fxString(RB, "Edit", ELLIPSIS), event -> onEditTransaction());
        var deleteMenuItem = menuItem(fxString(RB, "Delete", ELLIPSIS), event -> onDeleteTransaction());
        var exportMenuItem = menuItem(fxString(RB, "menu.Context.Export"), event -> onExportTransactions());
        var detailsMenuItem = menuItem(fxString(RB, "menu.item.details"), event -> onTransactionDetails());
        var checkMenuItem = menuItem(fxString(RB, "menu.item.check"), event -> onCheckTransactions(true));
        var uncheckMenuItem = menuItem(fxString(RB, "menu.item.uncheck"), event -> onCheckTransactions(false));

        editMenuItem.disableProperty().bind(getSelectionModel().selectedItemProperty().isNull());
        deleteMenuItem.disableProperty().bind(getSelectionModel().selectedItemProperty().isNull());
        detailsMenuItem.disableProperty().bind(getSelectionModel().selectedItemProperty().isNull());

        var ctxMenu = new ContextMenu();

        if (mode == Mode.SUMMARY) {
            ctxMenu.getItems().addAll(
                newMenuItem
            );
        }

        ctxMenu.getItems().addAll(
            editMenuItem,
            new SeparatorMenuItem()
        );

        if (mode == Mode.SUMMARY || mode == Mode.QUERY || mode == Mode.ACCOUNT) {
            ctxMenu.getItems().addAll(
                deleteMenuItem,
                new SeparatorMenuItem()
            );
        }

        if (mode != Mode.STATEMENT) {
            ctxMenu.getItems().addAll(
                exportMenuItem,
                new SeparatorMenuItem(),
                detailsMenuItem,
                new SeparatorMenuItem()
            );
        }

        ctxMenu.getItems().addAll(
            checkMenuItem,
            uncheckMenuItem
        );

        setContextMenu(ctxMenu);
    }

    Predicate<Transaction> getTransactionFilter() {
        return transactionPredicateProperty.get();
    }

    void setTransactionFilter(Predicate<Transaction> filter) {
        transactionPredicateProperty.set(filter.and(t -> t.parentUuid() == null));
        listSizeProperty.set(filteredList.size());
    }

    Optional<Transaction> getSelectedTransaction() {
        return Optional.ofNullable(getSelectionModel().getSelectedItem());
    }

    private void onCheckTransaction(List<Transaction> t, boolean checked) {
        checkTransactionConsumer.accept(t, checked);
    }

    void setOnCheckTransaction(BiConsumer<List<Transaction>, Boolean> c) {
        checkTransactionConsumer = c;
    }

    private void onExportTransactions() {
        var toExport = getSelectionModel().getSelectedItems();
        if (toExport.isEmpty()) {
            return;
        }

        var fileChooser = new FileChooser();
        fileChooser.setTitle("Export to file");
        Options.getLastExportDir().ifPresent(fileChooser::setInitialDirectory);
        fileChooser.getExtensionFilters().addAll(FILTER_XML_FILES, FILTER_ALL_FILES);

        var selected = fileChooser.showSaveDialog(null);
        if (selected == null) {
            return;
        }

        CompletableFuture.runAsync(() -> {
            try (var out = new FileOutputStream(selected)) {
                new Export().withTransactions(toExport, true)
                    .doExport(out);
                Options.setLastExportDir(selected.getParent());
            } catch (IOException ex) {
                throw new UncheckedIOException(ex);
            }
        });
    }

    private void redraw() {
        getColumns().get(0).setVisible(false);
        getColumns().get(0).setVisible(true);
    }

    void onTransactionDetails() {
        getSelectedTransaction().ifPresent(t -> {
            var childTransactions = cache().getTransactionDetails(t);

            if (mode == Mode.SUMMARY) {
                if (transactionDetailsCallback == null) {
                    return;
                }
                new TransactionDetailsDialog(childTransactions, t.amount(), false)
                    .showAndWait()
                    .ifPresent(list -> transactionDetailsCallback.handleTransactionDetails(t, list));
            } else {
                new TransactionDetailsDialog(childTransactions, BigDecimal.ZERO, true).showAndWait();
            }
        });
    }

    void onCheckTransactions(boolean check) {
        var selection = getCurrentSelection();
        var process = getSelectionModel().getSelectedItems().stream()
            .filter(t -> t.checked() != check)
            .collect(Collectors.toList());

        onCheckTransaction(process, check);
        restoreSelection(selection);
    }

    void onNewTransaction() {
        new TransactionDialog(null, options().getDialogCssFileUrl(), cache()).showAndWait().ifPresent(
            builder -> transactionAddedCallback.accept(getDao().insertTransaction(builder))
        );
    }

    void onEditTransaction() {
        var selection = getCurrentSelection();
        getSelectedTransaction()
            .flatMap(selected -> new TransactionDialog(null, options().getDialogCssFileUrl(), selected, cache()).showAndWait())
            .ifPresent(builder -> transactionUpdatedCallback.accept(getDao().updateTransaction(builder)));
        restoreSelection(selection);
    }

    void onDeleteTransaction() {
        getSelectedTransaction().ifPresent(transaction ->
            new Alert(Alert.AlertType.CONFIRMATION, "Are you sure to delete this transaction?")
                .showAndWait()
                .ifPresent(r -> {
                    if (r == ButtonType.OK) {
                        getDao().deleteTransaction(transaction);
                    }
                }));
    }

    boolean checkFocus() {
        return isFocused();
    }

    public ObservableList<Transaction> selectedTransactions() {
        return getSelectionModel().getSelectedItems();
    }

    private List<UUID> getCurrentSelection() {
        return getSelectionModel().getSelectedItems().stream().map(Transaction::uuid).collect(Collectors.toList());
    }

    /**
     * This method checks if table has any items with uuid from the list and selects all that exists.
     *
     * @param selection uuid of items to be selected
     */
    private void restoreSelection(List<UUID> selection) {
        Platform.runLater(() -> {
            getSelectionModel().clearSelection();
            for (var uuid : selection) {
                getItems().stream()
                    .filter(t -> t.uuid().equals(uuid))
                    .findAny()
                    .ifPresent(t -> getSelectionModel().select(t));
            }
        });
    }
}

File: src/main/kotlin/org/panteleyev/money/TransactionTableView.kt
/*
 * Copyright (c) 2017, Petr Panteleyev <petr@panteleyev.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package org.panteleyev.money

import javafx.application.Platform
import javafx.beans.property.ReadOnlyIntegerProperty
import javafx.beans.property.ReadOnlyObjectWrapper
import javafx.beans.property.SimpleIntegerProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.collections.MapChangeListener
import javafx.scene.control.CheckBox
import javafx.scene.control.ContextMenu
import javafx.scene.control.MenuItem
import javafx.scene.control.SelectionMode
import javafx.scene.control.SeparatorMenuItem
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.stage.FileChooser
import org.panteleyev.money.cells.TransactionContactCell
import org.panteleyev.money.cells.TransactionCreditedAccountCell
import org.panteleyev.money.cells.TransactionDayCell
import org.panteleyev.money.cells.TransactionDebitedAccountCell
import org.panteleyev.money.cells.TransactionRow
import org.panteleyev.money.cells.TransactionSumCell
import org.panteleyev.money.cells.TransactionTypeCell
import org.panteleyev.money.persistence.MoneyDAO
import org.panteleyev.money.persistence.SplitTransaction
import org.panteleyev.money.persistence.Transaction
import org.panteleyev.money.persistence.TransactionGroup
import org.panteleyev.money.xml.Export
import java.io.FileOutputStream
import java.util.ArrayList
import java.util.Arrays
import java.util.Comparator
import java.util.ResourceBundle
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.function.Predicate

internal class TransactionTableView(fullDate: Boolean) : TableView<Transaction>() {

    private var addGroupConsumer: (TransactionGroup, List<Transaction>) -> Unit = { _, _ -> }
    private var ungroupConsumer: (List<Transaction>) -> Unit = { }
    private var checkTransactionConsumer: (List<Transaction>, Boolean) -> Unit = { _, _ -> }

    // Menu items
    private val ctxGroupMenuItem = MenuItem(BUNDLE.getString("menu.Edit.Group"))
    private val ctxUngroupMenuItem = MenuItem(BUNDLE.getString("menu.Edit.Ungroup"))

    // Columns
    val dayColumn: TableColumn<Transaction, Transaction>

    // Transaction filter
    private var transactionFilter = Predicate<Transaction> { false }

    // List size property
    private val listSizeProperty = SimpleIntegerProperty(0)

    init {
        setRowFactory { TransactionRow() }
        selectionModel.selectionMode = SelectionMode.MULTIPLE

        dayColumn = TableColumn<Transaction, Transaction>(BUNDLE.getString("column.Day"))
        dayColumn.setCellValueFactory { p: TableColumn.CellDataFeatures<Transaction, Transaction> -> SimpleObjectProperty(p.value) }
        dayColumn.setCellFactory { TransactionDayCell(fullDate) }
        dayColumn.isSortable = true

        if (fullDate) {
            dayColumn.comparatorProperty().set(Comparator<Transaction> { o1, o2 ->
                var res = o1.year - o2.year
                if (res != 0) {
                    res
                } else {
                    res = o1.month - o2.month
                    if (res != 0) {
                        res
                    } else {
                        o1.day - o2.day
                    }
                }
            })
        } else {
            dayColumn.comparatorProperty().set(Comparator.comparingInt<Transaction>({ it.day }))
        }

        sortPolicy = TransactionTableSortPolicy()

        val typeColumn = TableColumn<Transaction, Transaction>(BUNDLE.getString("column.Type"))
        typeColumn.setCellValueFactory { p: TableColumn.CellDataFeatures<Transaction, Transaction> ->
            val v = if (p.value is SplitTransaction || p.value.groupId == 0)
                p.value
            else
                null

            ReadOnlyObjectWrapper<Transaction>(v)
        }
        typeColumn.setCellFactory { TransactionTypeCell() }

        typeColumn.comparator = Comparator.comparingInt { t: Transaction -> t.transactionType.id }
                .thenComparing(dayColumn.comparator)
        typeColumn.isSortable = true

        val accountFromColumn = TableColumn<Transaction, Transaction>(BUNDLE.getString("column.Account.Debited"))
        accountFromColumn.setCellFactory { TransactionDebitedAccountCell() }
        accountFromColumn.setCellValueFactory { p: TableColumn.CellDataFeatures<Transaction, Transaction> ->
            val v = if (p.value is SplitTransaction || p.value.groupId == 0)
                p.value
            else
                null

            ReadOnlyObjectWrapper<Transaction>(v)
        }
        accountFromColumn.comparator = Comparator.comparingInt<Transaction>({ it.accountDebitedId })
                .thenComparing(dayColumn.comparator)
        typeColumn.isSortable = true

        val accountCreditedColumn = TableColumn<Transaction, Transaction>(BUNDLE.getString("column.Account.Credited"))
        accountCreditedColumn.setCellFactory { TransactionCreditedAccountCell() }
        accountCreditedColumn.setCellValueFactory {
            p: TableColumn.CellDataFeatures<Transaction, Transaction> ->
            ReadOnlyObjectWrapper(p.value)
        }
        accountCreditedColumn.isSortable = false

        val contactColumn = TableColumn<Transaction, Transaction>(BUNDLE.getString("column.Payer.Payee"))
        contactColumn.setCellFactory { TransactionContactCell() }
        contactColumn.setCellValueFactory {
            p: TableColumn.CellDataFeatures<Transaction, Transaction> ->
            ReadOnlyObjectWrapper(p.value)
        }
        contactColumn.isSortable = false

        val commentColumn = TableColumn<Transaction, String>(BUNDLE.getString("column.Comment"))
        commentColumn.setCellValueFactory {
            p: TableColumn.CellDataFeatures<Transaction, String> ->
            ReadOnlyObjectWrapper(p.value.comment)
        }
        commentColumn.isSortable = false

        val sumColumn = TableColumn<Transaction, Transaction>(BUNDLE.getString("column.Sum"))
        sumColumn.setCellValueFactory {
            p: TableColumn.CellDataFeatures<Transaction, Transaction> ->
            ReadOnlyObjectWrapper(p.value)
        }
        sumColumn.setCellFactory { TransactionSumCell() }
        sumColumn.setComparator({ t1, t2 -> t1.signedAmount.compareTo(t2.signedAmount) })
        sumColumn.isSortable = true

        val approvedColumn = TableColumn<Transaction, CheckBox>("")

        approvedColumn.setCellValueFactory { p ->
            val cb = CheckBox()

            val value = p.value
            if (value is SplitTransaction) {
                val transactions = getTransactionsByGroup(value.groupId)
                cb.isSelected = transactions.all { it.checked }
                cb.setOnAction { onCheckTransaction(transactions, cb.isSelected) }
            } else {
                cb.isDisable = value.groupId != 0
                cb.isSelected = value.checked
                cb.setOnAction { onCheckTransaction(listOf(value), cb.isSelected) }
            }

            ReadOnlyObjectWrapper(cb)
        }
        approvedColumn.isSortable = false

        columns.setAll(Arrays.asList(dayColumn,
                typeColumn,
                accountFromColumn,
                accountCreditedColumn,
                contactColumn,
                commentColumn,
                sumColumn,
                approvedColumn
        ))

        sortOrder.add(dayColumn)
        dayColumn.sortType = TableColumn.SortType.ASCENDING

        // Column width. Temporary solution, there should be a better option.
        dayColumn.prefWidthProperty().bind(widthProperty().subtract(20).multiply(0.05))
        typeColumn.prefWidthProperty().bind(widthProperty().subtract(20).multiply(0.1))
        accountFromColumn.prefWidthProperty().bind(widthProperty().subtract(20).multiply(0.1))
        accountCreditedColumn.prefWidthProperty().bind(widthProperty().subtract(20).multiply(0.1))
        contactColumn.prefWidthProperty().bind(widthProperty().subtract(20).multiply(0.2))
        commentColumn.prefWidthProperty().bind(widthProperty().subtract(20).multiply(0.35))
        sumColumn.prefWidthProperty().bind(widthProperty().subtract(20).multiply(0.05))
        approvedColumn.prefWidthProperty().bind(widthProperty().subtract(20).multiply(0.05))

        selectionModel.selectionMode = SelectionMode.MULTIPLE
        val exportMenuItem = MenuItem(BUNDLE.getString("menu.Context.Export")).apply {
            setOnAction { onExportTransactions() }
        }

        contextMenu = ContextMenu(exportMenuItem).apply {
            setOnShowing { onShowingContextMenu() }
        }

        // No full date means we are in the transaction editing enabled mode
        if (!fullDate) {
            // Context menu
            ctxGroupMenuItem.setOnAction { onGroup() }
            ctxUngroupMenuItem.setOnAction { onUngroup() }

            contextMenu.items.addAll(
                    SeparatorMenuItem(),
                    ctxGroupMenuItem,
                    ctxUngroupMenuItem
            )
        }

        with(MoneyDAO) {
            transactions().addListener(MapChangeListener<Int, Transaction> { transactionListener(it) })
            transactionGroups().addListener(MapChangeListener<Int, TransactionGroup> { transactionGroupListener(it) })
        }
    }

    fun listSizeProperty(): ReadOnlyIntegerProperty {
        return listSizeProperty
    }

    private fun onShowingContextMenu() {
        val disableGroup: Boolean
        val disableUngroup: Boolean

        val selectedItems = selectionModel.selectedItems
        if (selectedItems.isEmpty()) {
            disableGroup = true
            disableUngroup = true
        } else {
            if (selectedItems.size == 1) {
                disableGroup = true
                val item = selectedItems[0]
                disableUngroup = item !is SplitTransaction
            } else {
                disableUngroup = true

                // Check if we don't have groups or group members selected
                val hasGroups = selectedItems.any { it.groupId != 0 }
                if (hasGroups) {
                    disableGroup = true
                } else {
                    // All transactions must have same day and debited account
                    val day = selectedItems[0].day
                    val debitedId = selectedItems[0].accountDebitedId

                    disableGroup = selectedItems.any { it.day != day || it.accountDebitedId != debitedId }
                }
            }
        }

        ctxGroupMenuItem.isDisable = disableGroup
        ctxUngroupMenuItem.isDisable = disableUngroup
    }

    fun clear() {
        items.clear()
    }

    private fun addRecords(transactions: List<Transaction>) {
        // Add transactions with no group
        transactions.filter { it.groupId == 0 }
                .forEach { items.add(it) }

        // Add groups
        val transactionsByGroup = transactions
                .filter { it.groupId != 0 }
                .groupBy { it.groupId }

        transactionsByGroup.keys.forEach { groupId ->
            val group = transactionsByGroup[groupId]
            if (group != null && !group.isEmpty()) {
                val split = SplitTransaction(groupId, group)
                items.add(split)
                items.addAll(group)
            }
        }

        listSizeProperty.set(items.size)
    }

    fun setTransactionFilter(filter: Predicate<Transaction>) {
        transactionFilter = filter

        selectionModel.clearSelection()
        clear()

        addRecords(MoneyDAO.getTransactions()
                .filter { filter.test(it) })
        sort()
    }

    val selectedTransactionCount: Int
        get() = selectionModel.selectedIndices.size

    val selectedTransaction: Transaction?
        get() {
            val transaction = selectionModel.selectedItem
            return if (transaction is SplitTransaction)
                null
            else
                transaction
        }

    // Method assumes that all checks are done in onShowingContextMenu method
    private fun onGroup() {
        // List of selected transactions
        val transactions = selectionModel.selectedItems

        val first = transactions[0]

        val tg = TransactionGroup(id = 0,
                day = first.day,
                month = first.month,
                year = first.year,
                expanded = false,
                guid = UUID.randomUUID().toString(),
                modified = System.currentTimeMillis()
        )

        addGroupConsumer(tg, ArrayList(transactions))
    }

    // Method assumes that all checks are done in onShowingContextMenu method
    private fun onUngroup() {
        ungroupConsumer(getTransactionsByGroup(selectionModel.selectedItem.groupId))
    }

    private fun onCheckTransaction(t: List<Transaction>, checked: Boolean) {
        checkTransactionConsumer(t, checked)
    }

    fun setOnAddGroup(bc: (TransactionGroup, List<Transaction>) -> Unit) {
        addGroupConsumer = bc
    }

    fun setOnDeleteGroup(c: (List<Transaction>) -> Unit) {
        ungroupConsumer = c
    }

    fun setOnCheckTransaction(c: (List<Transaction>, Boolean) -> Unit) {
        checkTransactionConsumer = c
    }

    private fun getTransactionsByGroup(groupId: Int): List<Transaction> {
        return items.filter { it.groupId == groupId && it !is SplitTransaction }
    }

    private fun onExportTransactions() {
        val toExport = selectionModel.selectedItems
                .filterNot { it is SplitTransaction }

        if (!toExport.isEmpty()) {
            val selected = FileChooser().apply {
                title = "Export to file"
                extensionFilters.addAll(
                        FileChooser.ExtensionFilter("XML Files", "*.xml"),
                        FileChooser.ExtensionFilter("All Files", "*.*")
                )
            }.showSaveDialog(null)

            selected?.let {
                CompletableFuture.runAsync {
                    FileOutputStream(selected).use {
                        Export().withTransactions(toExport, withDeps = true)
                                .export(it)
                    }
                }
            }
        }
    }

    private fun transactionListener(change: MapChangeListener.Change<out Int, out Transaction>) {
        val added = if (change.wasAdded() && transactionFilter.test(change.valueAdded))
            change.valueAdded
        else
            null

        val index = if (change.wasRemoved())
            items.indexOf(change.valueRemoved)
        else
            -1

        if (index != -1) {
            if (added != null) {
                items[index] = added
            } else {
                items.removeAt(index)
            }
        } else {
            if (added != null) {
                items.add(added)
            }
        }

        listSizeProperty.set(items.size)

        Platform.runLater {
            sort()
            if (added != null) {
                selectionModel.clearSelection()
                selectionModel.select(added)
            }
        }
    }

    private fun transactionGroupListener(change: MapChangeListener.Change<out Int, out TransactionGroup>) {
        if (change.wasRemoved()) {
            // find appropriate group
            items.filter { it.groupId == change.valueRemoved.id && it is SplitTransaction }
                    .firstOrNull()
                    ?.let { items.remove(it) }
        }

        if (change.wasAdded()) {
            // check if this list has transactions from this group
            val group = items.filter { it.groupId == change.valueAdded.id && it !is SplitTransaction }
            if (!group.isEmpty()) {
                val split = SplitTransaction(change.valueAdded.id, group)
                items.add(split)
                Platform.runLater { this.sort() }
            }
        }

        listSizeProperty.set(items.size)
    }

    companion object {
        private val BUNDLE = ResourceBundle.getBundle(MainWindowController.UI_BUNDLE_PATH)
    }
}
